### PR TITLE
Create deployment pipeline

### PR DESCRIPTION
**Pull request by kylelaker-ccpo**
This adopts a pipeline to deploy changes to non-sandbox environments and
solidifies the idea of what is in fact a "sandbox" environment by simply
leveraging a new context value. The pipeline is implemented using the
CDK's native CDK Pipelines construct library (`aws-cdk-lib.pipelines`).
This uses CodePipeline under the hood to perform the deployment.

Through this, IAM is significantly refactored to remove no-longer-needed
values and permissions (and also adds permissions to view the pipeline)
and moves the "base infrastructure" to be managed by the pipeline. This
sort of starts to treat sandbox environments as a special case rather
than a "real" environment; this more accurately represents how our
environments actually look (where sandbox is the only "weird" account).

We actually remove quite a bit of code here. And a lot of the "new" code
is actually just code that has been moved from other places. For the
most part, the Pipeline stack itself and the `bin/` folder are the only
places that had much net-new code besides the README, which has been
entirely rewritten from the ground up. The changes there should likely
be reviewed carefully. Additionally, the steps developers need to take
to test their environments locally have changed; the details for those
changes are in the new README.

A temporary workaround has been put in place to ensure that we don't
break access to API Gateway from ServiceNow or the internet. We have
more work to do for network transit before we can make that work
perfectly.

The changes here also mean we no longer have custom values in our
`.secrets.baseline` (for now).

Implementation details for how we create and manage the GitHub PAT will
not be covered in this repository and (for now) we will keep those
details internal.

Ticket: AT-7204

### CODE REVIEW

File: lib/atat-pipeline-stack.ts
Comment (juliusfitzhugh-ccpo) [inline lib/atat-pipeline-stack.ts:27]: Does a `net` stack still get created if `props.vpcCidr` is `undefined`? It looks like a `vpcCidr` is not being supplied for the `app`, but I guess a default value might be provided by AWS. 

Also, do we still plan to have a shared vpc in sandbox, since we were having issues previously with running out of vpcs?

File: lib/atat-pipeline-stack.ts
Comment (juliusfitzhugh-ccpo) [inline lib/atat-pipeline-stack.ts:32]: I'm still wrapping my head around the infrastructure tests, but do we need new tests to go along with the new pipeline infrastructure?

File: cdk.json
Comment (juliusfitzhugh-ccpo) [inline cdk.json:30]: I don't think it makes a difference, but thought I'd ask. On line 14 is `yarn.lock` are we using `yarn` somewhere?  Also, on line 16 is `test`, we are no longer using a `test` folder anymore, so `test` is not needed is it?

**Review by juliusfitzhugh-ccpo — COMMENTED** [review]
This looks good and I'm excited to see this in action. I asked some questions, but they are more geared towards making sure I understand things. 

Once the pipeline is deployed, then all the following merges into `develop` will keep the application updated without the need to manually deploy right? Or will we have another step of approval before that happens?

File: lib/atat-pipeline-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-pipeline-stack.ts:27]: It _should_ it will just use the default CIDR range. But good catch! I'll make sure we have a way to propagate that down to the net stack. The App stack shouldn't need it.

File: lib/atat-pipeline-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-pipeline-stack.ts:32]: Yeah, we do. That's why Sonar is grumpy right now :)

File: cdk.json
Comment (kylelaker-ccpo) [reply cdk.json:30]: Oh yeah, everything in that block above is part of the default template from `cdk init`. I think it's for the `cdk watch` command which we don't really use (or at least I don't _think_ any of us do but I'd love to see someone try it). So are the default feature flags in the context.

File: bin/atat-web-api.ts
Comment (kylelaker-ccpo) [reply bin/atat-web-api.ts:33]: Yep! That gets covered in the second commit message but I should have elevated that info into the PR body text as well. I'll work on updating the PR in a bit:

> This configures the application pipeline to deploy a "full" environment. It also allows for building "sandboxes" disconnected from a VPC and the rest of the environment. There will still be a "Sandbox" environment to build some guardrails for IAM (and the pipeline itself for that) but it's really no big deal (hopefully). Ideally, the only developer-facing change to deploy a sandbox environment would be the need to now add `-c atat:sandbox=true` to CLI invocations for diff/deploy.

File: lib/atat-pipeline-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-pipeline-stack.ts:27]: Also... as far as VPC sharing, I was almost thinking that we just wouldn't care about the VPC in Sandbox environments? And make it a dev environment problem? And just rely on the "default" networking capabilities of the various services/resources in sandbox? Maybe that's a little too ambitious.

Because I _do_ still plan to deploy a `SandboxPipeline` (so that it can manage IAM), there actually _would_ be a Sandbox VPC (and IAM and API) that would track `develop` but could be configured to track a feature branch if that'd be useful for a ticket or for testing.

Comment (juliusfitzhugh-ccpo) [reply lib/atat-pipeline-stack.ts:27]: Yeah, I don't think we care too much about the VPC in sandbox. Plus, we are doing less in AWS, so maybe things will be better. The VPCs were just resource intensive before

Cool, that makes sense (our IAM is every evolving). It might be nice to have that flexibility to track a feature branch for testing purposes, but I'll leave that up to you.

File: cdk.json
Comment (juliusfitzhugh-ccpo) [reply cdk.json:30]: Oh okay. Yeah, it does look like it's for the `cdk watch` which can also use `hot-swap` that you mentioned previously. This will help speed up deployment, so if I get a chance I'll try to use it.

File: bin/atat-web-api.ts
Comment (juliusfitzhugh-ccpo) [reply bin/atat-web-api.ts:33]: Got it, I didn't go over each commit this time so I missed it.

File: lib/atat-pipeline-stack.ts
Comment (zachclark-ccpo) [inline lib/atat-pipeline-stack.ts:36]: Since we are using s3, should we include:
`selfMutation: true` ?
v2 docs: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines-readme.html#s3-error-access-denied

ShellStep docs explicitly setting `selfMutation` to `true` or `false`: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.ShellStep.html#example

File: lib/atat-web-api-stack.ts
Comment (zachclark-ccpo) [inline lib/atat-web-api-stack.ts:29]: Should we throw an error here if `network` isn't provided?

File: lib/atat-pipeline-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-pipeline-stack.ts:36]: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#selfmutation

> selfMutation?
> Type: boolean (optional, default: true)

We should be good to go here.

File: lib/atat-net-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-net-stack.ts:99]: I don't know if I love it... I am playing around with this. We'll see if I end up reverting it but I like that it means we can map back to services.

File: lib/atat-pipeline-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-pipeline-stack.ts:39]: So right now, as I am developing, it's me/my account. I am going try to see if we can just provide an empty/invalid PAT since the repo is public. But otherwise, this is a conversation we'll need to come back to.

File: lib/atat-web-api-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-web-api-stack.ts:29]: We shouldn't. The better solution if we wanted to require `network` would to be to just... require `network` :) See https://github.com/dod-ccpo/atat-web-api/pull/739#discussion_r844507277 for more information.

File: lib/atat-pipeline-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-pipeline-stack.ts:40]: Okay so "use a fake token" didnt work... we will have to use a real one. Let's sync up on the best way to handle that.

**Review by zachclark-ccpo — CHANGES_REQUESTED** [review]
Please update the README with deployment instructions

File: lib/atat-iam-stack.ts
Comment (jeffsegal-ccpo) [inline lib/atat-iam-stack.ts:95]: Assuming I'm tracking this correctly, it's really nice to let CodePipeline handle some of this complexity.

File: README.md
Comment (jeffsegal-ccpo) [inline README.md:11]: Shall we define the acronym here? :)

**Review by jeffsegal-ccpo — APPROVED** [review]
README looks great. Excited to be moving in a more formal direction for CI/CD while still maintaining the niceties of the sandbox environments for devs.

File: lib/atat-iam-stack.ts
Comment (kylelaker-ccpo) [reply lib/atat-iam-stack.ts:95]: In this case, it's actually CDKv2. This has been dead code for a few weeks now. The specifics for this role are now defined as part of the CDK Bootstrap stack. That becomes more useful especially when we move towards considering a cross-account setup.

File: README.md
Comment (kylelaker-ccpo) [reply README.md]: Done! Had to reflow the text a bit so the commit isn't quite as clean as just adding
the word.

**Pull request merged**
